Fixed #28144 -- Added FileSystemStorage.OS_OPEN_FLAGS to allow customization.

jonprindiville opened the pull request:
Moves the flags passed to `os.open` in `FileSystemStorage._save` to a class attribute, `FileSystemStorage. OS_OPEN_FLAGS`. Enables subclassers to control the flags.

See [ticket 28144](https://code.djangoproject.com/ticket/28144) for details.

timgraham commented on tests/file_storage/tests.py:
```diff
@@ -290,6 +290,42 @@ def test_file_save_with_path(self):
+        stored_name_1 = self.storage.save(name, f_1)
+        self.assertEqual(stored_name_1, name)
+        self.assertTrue(self.storage.exists(stored_name_1))
+        self.assertTrue(
```
Use single lines for all these asserts -- we allow up to 119 characters when it improves readability.

timgraham commented on tests/file_storage/tests.py:
```diff
@@ -566,6 +602,53 @@ def test_custom_get_available_name(self):
+class OverwritingStorageTests(FileStorageTests):
+    storage_class = OverwritingStorage
+
+    def test_save_overwrite_behaviour(self):
```
use US spelling (behavior)

jonprindiville commented:
> I'm not sure if the first test is working as expected. I don't see it fail if s.O_EXCL is commented out. Is a different condition being tested?

Ah, `FileStorageTests.test_save_overwrite_behavior`? Unless there were concurrent attempts to write to that file name, I wouldn't expect it to fail if _only_ the `O_EXCL` bit were commented out.

From what I can tell, `FileSystemStorage` has two mechanisms it uses to try to avoid clobbering an existing file:
- `get_available_name` tries to return a name that is not currently in use
- `_save` uses the `O_EXCL` flag

The `O_EXCL` flag seems to be used just in case the name that _was not in use_ when `get_available_name` checked _is now in use_ at the time of the call out to the file system. There is [a comment referencing this race condition](https://github.com/django/django/pull/8476/files#diff-87c0869f58253f571c08ccf0fc5c7465R245) in `FileSystemStorage._save`.

You could probably cause `FileSystemStorage` to clobber files by commenting out the `O_EXCL` bit, but only by doing concurrent writes to the same filename (and getting lucky).

I'm not sure about testing a race condition like that in a unit test. _Should_ that race condition be tested here? Thoughts on how to reproduce it reliably?

---

The reason I added that `FileStorageTests.test_save_overwrite_behavior` test was because the subsequent commit (0772748601de59262b04523f1a0cbad245c8ed33) adds a new `allow_overwrite` kwarg to `_save`  and I wanted to make sure the behaviour of `FileSystemStorage` would not change. Misguided?

timgraham commented:
I was trying to make sure that the new test doesn't duplicate some existing test. I couldn't figure out what change to make in `_save()` that would cause the new test to fail.

theodesp commented:
None of the links you mentioned offer a thread safe solution as nothing stops a thread reading a file and another overwriting it on the same moment. The tests you added are also look duplicated.

I also don't like the usage of a boolean parameter as it adds complexity that is reflected in your tests.

I think you need just to provide a wrapper for safe and concurrent open as this overwrite depends on various os system parameters which you cannot handle safely.

jonprindiville commented:
@theodesp a few followup questions on your comments:

> None of the links you mentioned offer a thread safe solution as nothing stops a thread reading a file and another overwriting it on the same moment.

Which links are you talking about here? I'm not trying to stop anybody overwriting anything. The goal of this PR is to make the existing `FileSystemStorage` more flexible such that an end-user can use it as a base class in their own implementation of an overwriting storage backend.

The behaviour of `FileSystemStorage` is not changing.

> The tests you added are also look duplicated.

My mistake -- can you point me at the existing tests I'm duplicating?

> I also don't like the usage of a boolean parameter as it adds complexity that is reflected in your tests.

Ah, this was modelled on [`file_move_safe`](https://github.com/django/django/blob/550cb3a365dee4edfdd1563224d5304de2a57fda/django/core/files/move.py#L31) which also takes an `allow_overwrite` boolean kwarg to [govern the usage of the `O_EXCL` flag](https://github.com/django/django/blob/550cb3a365dee4edfdd1563224d5304de2a57fda/django/core/files/move.py#L57-L58). Not sure if Tim or other Django-folk take exception to this decision.

> I think you need just to provide a wrapper for safe and concurrent open, as this overwrite depends on various OS system parameters which you cannot handle safely.

Can you expand on this? Not sure what you mean

theodesp commented:
That's all reflected on the ticket https://code.djangoproject.com/ticket/28144 where you mention about sneakiness.

As for the test I just believe that the 2 test cases are the same [here](https://github.com/django/django/pull/8476/files#diff-1b22eaecd077bb34826c543d072a5448R293) and [here](https://github.com/django/django/pull/8476/files#diff-1b22eaecd077bb34826c543d072a5448R610)

For the usage of a boolean parameter maybe I'm grumpy and I don't like violating the single responsibility principle. But I ultimately believe that this leads to a code smell, which would lead to unmaintainable code that is difficult to understand and that has a lower risk of being easily reused. 

For example reading only this line here:
```python
flags = (os.O_WRONLY | os.O_CREAT |(os.O_EXCL if not allow_overwrite else 0) | getattr(os, 'O_BINARY', 0))
```
is a very sneaky way to introduce complexity as the boolean switch is hidden between 2 flags.

I'm sure that everyone will agree to this as one of the Python mantras is simple is better than complex.

As for a safer or concurrent open maybe I was thinking how the `file_move_safe` works as you can read in the code there are several try/catch scenarios and each one of them explains why it happens.

In general, I don't want to do bikeshedding here. I just went through the code changes as a new contributor and tried to understand whats going on. My main concern here is that while reading the changes it feels there is something wrong here, that's why I commented. What do you think?

jonprindiville commented:
@theodesp 

> None of the links [on the ticket] offer a thread safe solution as nothing stops a thread reading a file and another overwriting it on the same moment.

Oh, I included those as examples of the sort of advice you'll find today if you search the internet for things like [`django overwrite storage`](https://www.google.com/search?q=django+overwrite+storage). As I commented in the ticket: I _don't_ think that these are good implementations of an overwriting storage backend (and have encountered this type of thing causing issues in production before.)

> As for the test I just believe that the 2 test cases are the same here and here

One of them checks that `FileSystemStorage` doesn't overwrite files (this is the existing behaviour I do not want to change). The second uses an [overwriting storage-backend based on `FileSystemStorage`](https://github.com/django/django/pull/8476/files#diff-1b22eaecd077bb34826c543d072a5448R599) and tests that it _is_ able to overwrite files.

The second test case is part of `OverwritingStorageTests` and overrides the first one (part of its base class, `FileStorageTests`.)

> As for a safer or concurrent open maybe I was thinking how the file_move_safe works as you can read in the code there are several try/catch scenarios and each one of them explains why it happens.

Can you point me at something specific that `file_move_safe` is doing that you think that `FileSystemStorage._save` should be doing?

There are three `try` blocks in `file_move_safe`:
- `try/catch` around copying metadata between two different filesystems
- `try/catch` around about deleting a file
- `try/finally` around some locking after the call to `open`

IMO these things do not need to be added to `FileSystemStorage._save`:
- it is dealing with only one filesystem
- it is not deleting anything
- it already contains a very similar bit of locking

jonprindiville commented:
Hey @timgraham, anything else that I can do here?

timmartin commented on django/core/files/storage.py:
```diff
@@ -217,7 +217,7 @@ def directory_permissions_mode(self):
         return File(open(self.path(name), mode))
 
-    def _save(self, name, content):
+    def _save(self, name, content, allow_overwrite=False):
```
The parameter to this method seems odd to me. As it stands, no caller is explicitly using the parameter in a call to the method and I can't see many legitimate uses for it in derived classes. I think you're effectively just using the default value as a space to store a class-scope variable. Would it be simpler to just assign a member on the class? Or on the instance, if you prefer.

jonprindiville replied:
>I think you're effectively just using the default value as a space to store a class-scope variable.

Yeah, I think that this is a fair assessment.

As a class-level attribute I think that [it would look fairly reasonable](https://github.com/jonprindiville/django/compare/ticket_28144...ticket_28144--class-attribute) but I wonder about naming it `ALLOW_OVERWRITE`. Really in order to do the overwriting you need to modify `get_available_name` as well (since it is called before `_save`.)

[Another sketch I did was to pull out the `which flags should we pass to os.open` into a method](https://github.com/jonprindiville/django/compare/ticket_28144...jonprindiville:ticket_28144--method-hook) which could be overridden by a subclasser. My example `OverwritingStorage` just omits the O_EXCL from its list but it could also call super and then mask out the O_EXCL.

You have thoughts on either of those approaches?

(You mention setting it on the instance as a possibility, as well: did you mean accepting an `allow_overwrite` kwarg in `FileSystemStorage.__init__` or just that init hardcode a `self._allow_overwrite = False`?)

timmartin replied:
Having an overridable method seems like the most orthodox OOP solution (it's what a Java programmer would do :-) ) but I'm not convinced it really gives a useful abstraction: by coincidence it's the right place to make this one change, but I'm not sure there's a useful class of future modifications it opens up, so it feels like overkill to me.

My thought with an instance variable was just to set it in the constructor in the base class, and overwrite it in the subclass constructor (not exposing it as a kwarg). I'm not sure there's any advantage to this; I think I was thinking about this because it's what I'd do in C++.

I don't have a particularly strong feeling on this. I think if I were writing it I'd go with the class-level attribute.

jonprindiville commented:
Hey @timmartin -- thanks for the feedback!

> I agree that it's sensible to test for regressions of the original behaviour of the code you touched in FileSystemStorage, but I don't think your test actually does that.

Ah -- you mean that the test I've added doesn't test all of the paths through FileSystemStorage functionality?

> make `get_available_name` use a deterministic counter and reset its counter between the two phases of the test you can achieve this without anything too horrendous.

Can you expand on this? You want to be able to make asserts on the number of calls to `get_available_name`?

Poking around in `file_storage.tests` a little further I see a couple relevant things:

- `FileSaveRaceConditionTest` uses a `ContentFile` subclass that's been hacked to return chunks very slowly (nice trick!) in order to have two threads writing to the same file name at the same time and asserts that one got a suffix -- maybe the test I added against FileSystemStorage is unnecessary in light of this.

- `FileStorageTests.test_makedirs_race_handling` and `.test_remove_race_handling` both mock out methods on `os` to be able to cover various branches of some functionality -- this technique could also be used to simulate the O_EXCL case in `._save` if someone thinks it's necessary.


> Also FWIW, I think before this gets merged I'd prefer to see your commit messages edited a bit. I like a detailed commit message, but I think the detailed discussion here is better appended to the ticket where anyone who really cares can look for it. This isn't necessarily Django project policy, just my personal opinion.

Haha, yeah -- I hear that. My tendency is too much information in commit messages. A lot of what's in there presently is also already on the ticket so I will clean all of that duplicated stuff out of there before this is ready for merge.

timmartin commented:
> Ah -- you mean that the test I've added doesn't test all of the paths through FileSystemStorage functionality?

I don't particularly mean that it doesn't test all the paths possible, I just don't think it hits the specific thing you've changed. Maybe I got that wrong, though. 

To be clear, my position here isn't that you have to cover this case. My thinking is that if you're going to write a test for this case, it should cover the thing that's changed. I'd be happy with just taking this test case out.

> > make get_available_name use a deterministic counter and reset its counter between the two phases of the test you can achieve this without anything too horrendous.
> Can you expand on this? You want to be able to make asserts on the number of calls to get_available_name?

What I was thinking of was that you could do something in the test instrumentation class like:

    class TestStorage(FileSystemStorage)
        def __init__(self):
            self.counter = 0
            #...

        def get_available_name(self, name):
            self.counter += 1
            return name + str(self.counter)

then in the test

        storage = TestStorage()
        storage.save("foo", contents)

        storage = TestStorage()
        # or just
        # storage.counter = 0 
        storage.save("foo", contents)

In the second run, `get_available_name` should return the name of a file that already exists ("foo1"), so `os.open()` will throw `FileExistsError`. However, crucially it won't fail the next time it is run, or you get an infinite loop.

The `FileSaveRaceConditionTest` is a neat way of testing, but it doesn't look to me like it necessarily hits the case where `get_available_name` returns an existing file, since it looks like the second thread could see the file as already existing in `get_available_name` and not get as far as trying to `os.open` the file. IIUC this is actually indeterminate in the test.

You're right that mocking would definitely work as a solution but personally I don't think it's a good balance in this case because it really only tests that the `O_EXCL` flag is being passed, and that's obvious by inspection. In general tests like this feel like they false positive more often than they detect regressions, because it's testing the API and not the desired behaviour. The obvious counter-argument is that `os.open` is never going to change its API, so maybe in this case it's not such a problem. I certainly wouldn't object to a mock test, it just seems unnecessary.

jonprindiville commented:
Earlier, I commented the following about my first run at a "class attribute" version of this PR:

> As a class-level attribute I think that it would look fairly reasonable but I wonder about naming it ALLOW_OVERWRITE. Really in order to do the overwriting you need to modify get_available_name as well (since it is called before _save.)

After sitting on this for a while: that was a bad name for that boolean. I thought about names for a while and then thought (as @theodesp mentioned earlier) maybe a boolean is just not a great idea here.

Stepping back: what I wanted was to be able to change that list of flags in my subclass, maybe we can just put those flags into a class attribute (as in most recent commit here, c1ae7a7.)

Thoughts?

---

> To be clear, my position here isn't that you have to cover this case. My thinking is that if you're going to write a test for this case, it should cover the thing that's changed. I'd be happy with just taking this test case out.

Yeah, I'm feeling that way as well, now that I am looking at _this_ iteration of the code. FileSystemStorage's functionality isn't changing here, I've removed that test.

---

> I like a detailed commit message, but I think the detailed discussion here is better appended to the ticket where anyone who really cares can look for it.

Removed stuff that is already discussed on the ticket.
